merkle_patricia_tree/merkle_tree/src/main.rs

```diff
@@ -0,0 +1,533 @@
+    offset2 += head;
+    len     -= head;
+    }
+
+    // At this point, the ranges are non-overlapping
+    // and the swap can be done in a naive way.
+    for ii in 0..len {
+    input.swap(offset1 + ii, offset2 + ii);
+    }
+
+    input.to_vec()
+}
+
+/// Copy bytes from one location in the input and overwrite them at another
+/// location in the input
+fn copy(input: &mut [u8]) -> Vec<u8> {
+    // Nothing to do on an empty input
+    if input.is_empty() {
+        return input.to_vec();
+    }
+
+    // Pick a source and destination for a copy
+    let src    = rand_offset(input);
+    let srcrem = input.len() - src;
+    let dst    = rand_offset(input);
+    let dstrem = input.len() - dst;
+
+    // Pick a random length up to the max for both offsets
+    let len = rand_number_in_range(1,core::cmp::min(srcrem, dstrem));
+
+    // Perform a copy inplace in the input
+    overwrite_inplace(input, src, len, dst)
+}
+
+/// Take the bytes from `source` for `len` bytes in the input, and copy
+/// them to `dest`
+fn overwrite_inplace(input: &mut [u8], source: usize, len: usize, dest: usize) -> Vec<u8> {
+
+    let mut new_input = input.to_owned();
+
+    // Nothing to do
+    if len == 0 || source == dest { return new_input.to_vec(); }
+
+    if source < dest {
+        // Copy forwards
+        for ii in 0..len {
+            new_input[dest + ii] = new_input[source + ii];
+        }
+
+        new_input.to_vec()
+    } else {
+        // Copy backwards
+        for ii in (0..len).rev() {
+            new_input[dest + ii] = new_input[source + ii];
+        }
+
+        new_input.to_vec()
+    }
+}
+
+/// Take one location of the input and splice it into another
+fn inter_splice(input: &mut [u8]) -> Vec<u8> {
+    // Nothing to do on an empty input
+    if input.is_empty() {
+        return input.to_vec();
+    }
+
+    // Pick a source and destination for an insertion
+    let src    = rand_offset(input);
+    let srcrem = input.len() - src;
+    let dst    = rand_offset(input);
+
+    // Pick a random length
+    let len = rand_number_in_range(1,srcrem);
+
+    // Perform an insertion inplace in the input
+    insert_inplace(input, src, len, dst)
+}
+
+/// Take the bytes from `source` for `len` bytes in the input, and insert
+/// a copy of them at `dest`
+fn insert_inplace(input: &mut [u8], source: usize, len: usize, dest: usize)-> Vec<u8> {
+    // Nothing to do
+    if len == 0 || source == dest { return input.to_vec(); }
+
+    // Cap the insertion to the max input size
+    let mut new_len = len;
+    if MAX_INPUT_SIZE < input.len() {
+        new_len = core::cmp::min(len, MAX_INPUT_SIZE);
+    } else {
+        new_len = core::cmp::min(len, MAX_INPUT_SIZE - input.len());
+    }
+
+    // Create an interator to splice into the input
+    let rep = core::iter::repeat(b'\0').take(new_len);
+
+    // Expand at `dest` with `rep`, making room for the copy
+    let mut input_vec = input.to_vec();
+    input_vec.splice(dest..dest, rep);
```

Comment:
`dest..dest` is an empty range. I think you meant to use an inclusive range (`dest..=dest`).

Reply:
fixed the comment

Reply:
done